### PR TITLE
Adding softmax cross entropy loss

### PR DESCRIPTION
**Pull request by lastmansleeping**

### CODE REVIEW

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py [hunk 1]
```diff
@@ -6,11 +6,10 @@
...
-class RankOneListNet(ListwiseLossBase):
+class SoftmaxCrossEntropy(ListwiseLossBase):
     def get_loss_fn(self, **kwargs):
         """
-        Define a masked rank 1 ListNet loss
-        Additionally can pass in record positions to handle positional bias
+        Define a softmax cross entropy loss
 
         Returns
         -------
```
Comment (Ullimague) [inline python/ml4ir/applications/ranking/model/losses/listwise_losses.py:15]: I don't think line 17 is valid anymore.

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:15]: My bad. Good catch!

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py [hunk 2]
```diff
@@ -33,18 +32,7 @@ def _loss_fn(y_true, y_pred):
...
-
-            # Reshape the tensors so that we sum the losses from each record
-            y_true = tf.expand_dims(tf.squeeze(y_true), axis=-1)
-            y_pred = tf.expand_dims(tf.squeeze(y_pred), axis=-1)
-
-            # Scale the sum of losses down by number of queries in the batch
-            return tf.math.divide(bce(y_true, y_pred), batch_size)
+            return cce(y_true, tf.math.multiply(y_pred, mask))
 
         return _loss_fn
```
Comment (Ullimague) [inline python/ml4ir/applications/ranking/model/losses/listwise_losses.py:38]: Now get_final_activation_op is part of another class. There is no breaking changes or tests about it?

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:38]: I am inheriting that class, right?

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:38]: Both the losses use the same activation. They just use a different method to compute the final loss.

Comment (Ullimague) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:38]: Oh I missed that. Do we want to have a common parent class for it? Or are we fine keeping it that way?

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:38]: I think it should be fine because Rank One ListNet loss is a variation of the softmax cross entropy loss, that's all.

**Review by Ullimague — APPROVED** [review]

File: python/ml4ir/applications/ranking/tests/test_losses.py [hunk 3]
```diff
@@ -62,3 +62,12 @@ def test_rank_one_listnet(self):
         loss = self.run_default_pipeline(loss_key="rank_one_listnet")
 
         assert np.isclose(loss, 2.00879, rtol=0.05)
+
+    def test_softmax_cross_entropy(self):
+        """
+        Test model training and evaluate Rank One ListNet loss
```
Comment (balikasg) [inline python/ml4ir/applications/ranking/tests/test_losses.py:68]: This docstring should be about softmax cross-entropy loss.

File: python/ml4ir/applications/ranking/tests/test_losses.py [hunk 4]
```diff
@@ -62,3 +62,12 @@ def test_rank_one_listnet(self):
...
+        """
+
+        loss = self.run_default_pipeline(loss_key="softmax_cross_entropy")
+
+        assert np.isclose(loss, 1.19955, rtol=0.05)
```
Comment (balikasg) [inline python/ml4ir/applications/ranking/tests/test_losses.py:73]: These tests, where we run the pipeline and get `np.isclose` to something, have value, in the sense that the pipeline runs. They boost coverage too! 
This is not the whole story: ideally we want to eval the actual code in `SoftmaxCrossEntropy`  and `get_loss_fn` where the tensor calculations happen. Would you consider adding such a "toy" test?

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py [hunk 5]
```diff
@@ -6,23 +6,22 @@
 from ml4ir.applications.ranking.model.losses.loss_base import ListwiseLossBase
 
 
-class RankOneListNet(ListwiseLossBase):
+class SoftmaxCrossEntropy(ListwiseLossBase):
```
Comment (balikasg) [inline python/ml4ir/applications/ranking/model/losses/listwise_losses.py:9]: Are there any references we can add? Not paragraphs but at least a pointer for someone to understand a bit more. For instance, to me the difference with `categorical_cross_entropy` is not evident (I am lacking ranking context). Is is suitable both for binary and more fine-grained relevance?

File: python/ml4ir/applications/ranking/model/losses/loss_factory.py [hunk 6]
```diff
@@ -26,5 +26,7 @@ def get_loss(loss_key, scoring_type) -> RelevanceLossBase:
         return pointwise_losses.SigmoidCrossEntropy(loss_key=loss_key, scoring_type=scoring_type)
     elif loss_key == LossKey.RANK_ONE_LISTNET:
         return listwise_losses.RankOneListNet(loss_key=loss_key, scoring_type=scoring_type)
+    elif loss_key == LossKey.SOFTMAX_CROSS_ENTROPY:
```
Comment (balikasg) [inline python/ml4ir/applications/ranking/model/losses/loss_factory.py:29]: the separation in poitwise_losses and listwise_losses reads nice here ;-)

**Review by balikasg — APPROVED** [review]
Few comments, but nothing that blocks this PR!

File: python/ml4ir/applications/ranking/tests/test_losses.py (hunk 4 above)
Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/tests/test_losses.py:73]: I agree. I am going to change all of them to just unit tests actually. It will help with memory leak as well. I think the current build is failing because of memory overflow.

Comment (balikasg) [reply python/ml4ir/applications/ranking/tests/test_losses.py:73]: Did you have memory issues when training actual models (not toy in tests)? I have been training for several days but without issues for instance (which is good). Perhaps I had more memory.. But I remember having issues with tests (locally)

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/tests/test_losses.py:73]: Not on my local box. I think it only breaks on the autobuilds because of the docker container hard limitation. Our local boxes have a lot of swap size for handling this (is my guess). 

The main issue is that in tf2, keras models are not releasing memory. So for a test like this, we create the pipeline many times (so we are creating and deleting keras models) and the model is not getting "garbage collected" fully. :|
